main/src/ca/uwaterloo/flix/language/phase/Kinder.scala

```diff
@@ -0,0 +1,1077 @@
+/*
+ * Copyright 2021 Matthew Lutze
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.CompilationError
+import ca.uwaterloo.flix.language.ast.{Kind, _}
+import ca.uwaterloo.flix.language.errors.KindError
+import ca.uwaterloo.flix.util.Validation.{ToFailure, ToSuccess, flatMapN, mapN, traverse}
+import ca.uwaterloo.flix.util.{InternalCompilerException, Validation}
+
+/**
+  * Attributes kinds to the types in the AST.
+  *
+  * For enums, classes, instances, and type aliases,
+  * kinds are either explicit or assumed to be Star.
+  *
+  * For defs,
+  * kinds are either explicit or are inferred from their
+  * use in the formal parameters, return type and effect, and type constraints.
+  *
+  * In inferring types, variable type constructors are assumed to have kind * -> * -> * -> ???.
```

Comment:
Could we add the rules from the ticket here? and update them with the latest developments?